sklearn/ensemble/_hist_gradient_boosting/grower.py

```diff
@@ -271,6 +322,12 @@ def _intilialize_root(self, gradients, hessians, hessians_are_constant):
         self.root.partition_start = 0
         self.root.partition_stop = n_samples
 
+        # Value hard-coded here. Sharing compute_node_value() between the
+        # splitting.pyx and this files creates strong Python interactions and
+        # slows down the code.
+        self.root.value = -sum_gradients / (
+            sum_hessians + self.l2_regularization + 1e-15)
```

Comment:
thanks for all the nice investigations. I like all the changes, but this one makes me quite uncomfortable.

If there's no other way for us to handle the performance issue, maybe add a comment in `compute_node_value` that here there's a hard coded version?

Reply:
Your comment made me realize that we don't actually need the root's value to be ever computed, so I just set it to 0 ;)

Reply:
So there's no situation where the tree will be only a root node?

Reply:
No, that's enforced by `max_depth` and `max_leaf_nodes`

Reply:
now I'm confused. They're `max_*` constraints, not `min_*`. What am I missing?

Reply:
We still have limits on their lower bounds:

```py
        if max_leaf_nodes is not None and max_leaf_nodes <= 1:
            raise ValueError('max_leaf_nodes={} should not be'
                             ' smaller than 2'.format(max_leaf_nodes))
        if max_depth is not None and max_depth < 1:
            raise ValueError('max_depth={} should not be'
                             ' smaller than 1'.format(max_depth))```

Reply:
In fact it's true that we can still have single-node trees.

But in this case, I think it's correct tot set their value to 0: we don't want them to have any impact on the prediction. I added a test in c945f8e

CC @ogrisel